### PR TITLE
fix(codegen): fix unicode escaping in es5

### PR DESCRIPTION
**Pull request by williamtetlow**
**Description:**
Fixes https://github.com/swc-project/swc/issues/3617

Babel has the logic `if target <= ES5 && original source string contains any unicode map all non-ascii to unicode`.

It's a bit implicit as the logic is across a few plugins and also through the options passed to `jsesc` (that depend on target and babel version) [but here babel sets](https://github.com/babel/babel/blob/0058b7fef4d028d2826645b28e3b448517cd979d/packages/babel-plugin-transform-literals/src/index.ts#L19.) `node.extra = undefined` if there's any unicode in string literal

[And here if](https://github.com/babel/babel/blob/df27d542ef3871c342b085e41206d21a86673c1d/packages/babel-generator/src/generators/types.ts#L204) `node.extra` exists it will return the string and otherwise process it with `jsesc` 

- I've replicated this logic in this PR.
- I've also added the logic for converting unicode characters `> 0xFFFF` to surrogate pairs to support ES5.
- I've updated the TSC test suite to match TSC and assert against unicode code points instead of characters

**BREAKING CHANGE:**

**Related issue (if exists):**
https://github.com/swc-project/swc/issues/3617

### CODE REVIEW

**Comment by williamtetlow** [comment]
Just posting these links here for future reference

https://mathiasbynens.be/notes/javascript-escapes#unicode

https://mathiasbynens.be/notes/javascript-encoding#surrogate-formulae

File: crates/swc_ecma_codegen/src/lib.rs [hunk 1]
```diff
@@ -3070,7 +3071,12 @@ fn unescape_tpl_lit(s: &str, is_synthesized: bool) -> String {
     result
 }
 
-fn escape_without_source(v: &str, target: EsVersion, single_quote: bool) -> String {
+fn escape_without_source(
+    v: &str,
+    target: EsVersion,
+    single_quote: bool,
+    emit_non_ascii_as_unicode: bool,
```
Comment (williamtetlow) [inline crates/swc_ecma_codegen/src/lib.rs:3078]: setting this to `true` is equivalent to disabling `jsesc` `minimal` option

https://github.com/mathiasbynens/jsesc#minimal

File: crates/swc_ecma_codegen/src/lib.rs [hunk 2]
```diff
@@ -3134,25 +3156,34 @@ fn escape_with_source(
...
     let orig = match orig {
         Ok(orig) => orig,
         Err(v) => {
-            return escape_without_source(s, target, single_quote.unwrap_or(false));
+            return escape_without_source(s, target, single_quote.unwrap_or(false), false);
         }
     };
 
+    if target <= EsVersion::Es5 {
+        let emit_non_ascii_as_unicode = UNICODE_CODEPOINT.is_match(&orig);
```
Comment (kdy1) [inline crates/swc_ecma_codegen/src/lib.rs:3175]: The only concern about this PR I have is the performance of this line.
Do we really need regex?

Comment (williamtetlow) [reply crates/swc_ecma_codegen/src/lib.rs]: @kdy1 I think that's my JS brain defaulting to regex 😛  Simple string contains is enough, updated in [4a11bab](https://github.com/swc-project/swc/pull/3636/commits/4a11bab72e265ab7cdeecb1746262a6c48b317fb)

File: crates/swc_ecma_codegen/Cargo.toml [hunk 3]
```diff
@@ -14,6 +14,7 @@ bitflags = "1"
 memchr = "2.4.1"
 num-bigint = {version = "0.2", features = ["serde"]}
 once_cell = "1.9.0"
+regex = "1"
```
Comment (kdy1) [inline crates/swc_ecma_codegen/Cargo.toml:17]: Can you remove this line?

**Review by kdy1 — CHANGES_REQUESTED** [review]
It's mostly LGTM.

---

swc-bump:
  - swc_ecma_codegen

File: crates/swc_ecma_codegen/Cargo.toml (hunk 3 above)
Comment (kdy1) [reply crates/swc_ecma_codegen/Cargo.toml:17]: Also, please revert `Cargo.lock`

Comment (williamtetlow) [reply crates/swc_ecma_codegen/Cargo.toml]: resolved in 0ef810b4a4d699c0ebf59a02d5cc139e0f41c062

**Review by kdy1 — APPROVED** [review]
Thank you!

---

swc-bump:
 - swc_ecma_codegen

**Pull request merged**